src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -34640,6 +34640,519 @@ .maxstack  2
+                    IL_001e:  ldloca.s   V_1
+                    IL_0020:  call       "bool MyCollection.Enumerator.MoveNext()"
+                    IL_0025:  brtrue.s   IL_000f
+                    IL_0027:  leave.s    IL_0037
+                  }
+                  finally
+                  {
+                    IL_0029:  ldloca.s   V_1
+                    IL_002b:  constrained. "MyCollection.Enumerator"
+                    IL_0031:  callvirt   "void System.IDisposable.Dispose()"
+                    IL_0036:  endfinally
+                  }
+                  IL_0037:  ldloc.0
+                  IL_0038:  ldarg.1
+                  IL_0039:  callvirt   "void System.Collections.Generic.List<int>.AddRange(System.Collections.Generic.IEnumerable<int>)"
+                  IL_003e:  ldloc.0
+                  IL_003f:  ret
+                }
+                """);
+        }
+
+        [Fact]
+        public void List_SingleSpread_CustomCollection_ICollectionAndStructEnumerator()
+        {
+            var source = """
+                using System.Collections;
+                using System.Collections.Generic;
+
+                class MyCollection(List<int> list) : ICollection<int>
+                {
+                    public int Count => list.Count;
+
+                    public bool IsReadOnly => false;
+
+                    public void Add(int item) => list.Add(item);
+
+                    public void Clear() => list.Clear();
+
+                    public bool Contains(int item) => list.Contains(item);
+
+                    public void CopyTo(int[] array, int arrayIndex) => list.CopyTo(array, arrayIndex);
+
+                    public bool Remove(int item) => list.Remove(item);
+
+                    public Enumerator GetEnumerator() => new(list.GetEnumerator());
+
+                    IEnumerator<int> IEnumerable<int>.GetEnumerator() => GetEnumerator();
+
+                    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+                    public struct Enumerator(List<int>.Enumerator enumerator) : IEnumerator<int>
+                    {
+                        public int Current => enumerator.Current;
+
+                        object IEnumerator.Current => Current;
+
+                        public bool MoveNext() => enumerator.MoveNext();
+
+                        public void Dispose() => enumerator.Dispose();
+
+                        public void Reset() { }
+                    }
+                }
+
+                class C
+                {
+                    static void Main()
+                    {
+                        M(new([1, 2, 3])).Report();
+                    }
+
+                    static List<int> M(MyCollection c) => [..c];
+                }
+                """;
+
+            var verifier = CompileAndVerify([source, s_collectionExtensionsWithSpan], expectedOutput: IncludeExpectedOutput("[1, 2, 3],"), targetFramework: TargetFramework.Net80, verify: Verification.Skipped);
```

Comment:
>var verifier = CompileAndVerify([source, s_collectionExtensionsWithSpan], expectedOutput: IncludeExpectedOutput("[1, 2, 3],"), targetFramework: TargetFramework.Net80, verify: Verification.Skipped); [](http://example.com/codeflow?start=12&length=197)

Consider also compiling and verifying IL for the default target framework. That should avoid the `CollectionsMarshal` optimization and use `AddRange()` instead. And it should be possible to test `expectedOutput` regardless of platform for that case. Please consider adding to this test rather than creating a separate test with duplicated source.
```
verifier = CompileAndVerify([source, s_collectionExtensions], expectedOutput: "[1, 2, 3],", verify: Verification.Skipped);
verifier.VerifyIL("C.M", ...);
```